models/repo_watch.go

```diff
@@ -4,42 +4,118 @@
 
 package models
 
-import "fmt"
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// RepoWatchMode specifies what kind of watch the user has on a repository
+type RepoWatchMode int8
+
+const (
+	// RepoWatchModeNone don't watch
+	RepoWatchModeNone RepoWatchMode = iota // 0
+	// RepoWatchModeNormal watch repository (from other sources)
+	RepoWatchModeNormal // 1
+	// RepoWatchModeDont explicit don't auto-watch
+	RepoWatchModeDont // 2
+	// RepoWatchModeAuto watch repository (from AutoWatchOnChanges)
+	RepoWatchModeAuto // 3
+)
 
 // Watch is connection request for receiving repository notification.
 type Watch struct {
-	ID     int64 `xorm:"pk autoincr"`
-	UserID int64 `xorm:"UNIQUE(watch)"`
-	RepoID int64 `xorm:"UNIQUE(watch)"`
+	ID     int64         `xorm:"pk autoincr"`
+	UserID int64         `xorm:"UNIQUE(watch)"`
+	RepoID int64         `xorm:"UNIQUE(watch)"`
+	Mode   RepoWatchMode `xorm:"SMALLINT NOT NULL DEFAULT 1"`
```

Comment:
We needs a migration from old data to new data. And why not use 0 for watch?

Reply:
This way I didn't need a migration; but perhaps it's neater having one. I'll do it tonight.

Reply:
@lunny Thinking about it a little more, if we make `RepoWatchModeNormal == 0`, what value  should I use for `RepoWatchModeNone` then? I've chosen `RepoWatchModeNone == 0` because that's the only value that is never created in the database: only the other values create a row in `watch`. I think it would be weird to have rows with values `0`, `2`, `3` and `4` but never `1`. Buy hey, that's the problem with enums anyway. `RepoWatchModeNone` should never be saved to the database; it wouldn't make sense.

Reply:
I've added the migration, but I'm not sure if it's really needed. The column seems to be created "automagically" by xorm.

Reply:
Yes it is created automatically but if there will be added migration in future that does use this column in any way, it will break upgrades from older versions as first migrations are run and only then xorm magic. This way migration could try to select from column that does not exist in database